alephium/lib/deployment.ts

```diff
@@ -0,0 +1,365 @@
+  const account = accounts[0]
+
+  const deployContract = async (contract: Contract, params: DeployContractParams): Promise<DeployContractResult> => {
+    // TODO: improve this after migrating to SDK
+    const bytecode = contract.buildByteCodeToDeploy(params.initialFields ? params.initialFields : {})
+    const codeHash = cryptojs.SHA256(bytecode).toString()
+    const previous = deployContractResults.get(contract.name)
+    const tokens = params.initialTokenAmounts ? getTokenRecord(params.initialTokenAmounts) : undefined
+    const issueTokenAmount: string | undefined = params.issueTokenAmount?.toString()
+    const needToDeploy = await needToDeployContract(
+      signer.provider,
+      previous,
+      params.initialAttoAlphAmount,
+      tokens,
+      issueTokenAmount,
+      codeHash
+    )
+    if (!needToDeploy) {
+      // we have checked in `needToDeployContract`
+      return previous!
+    }
+    const result = await contract.transactionForDeployment(signer, params)
+    await signer.submitTransaction(result.unsignedTx)
+    const confirmed = await waitTxConfirmed(signer.provider, result.txId, network.confirmations)
+    const deployContractResult: DeployContractResult = {
+      fromGroup: result.fromGroup,
+      toGroup: result.toGroup,
+      txId: result.txId,
+      blockHash: confirmed.blockHash,
+      contractId: result.contractId,
+      contractAddress: result.contractAddress,
+      codeHash: codeHash,
+      attoAlphAmount: params.initialAttoAlphAmount,
+      tokens: tokens,
+      issueTokenAmount: issueTokenAmount
+    }
+    deployContractResults.set(contract.name, deployContractResult)
+    return deployContractResult
+  }
+
+  const runScript = async (script: Script, params: RunScriptParams, taskName?: string): Promise<RunScriptResult> => {
+    const bytecode = script.buildByteCodeToDeploy(params.initialFields ? params.initialFields : {})
+    const codeHash = cryptojs.SHA256(bytecode).toString()
+    const key = taskName ? taskName : script.name
+    const previous = runScriptResults.get(key)
+    const tokens = params.tokens ? getTokenRecord(params.tokens) : undefined
+    const needToRun = await needToRunScript(signer.provider, previous, params.attoAlphAmount?.toString(), tokens, codeHash)
+    if (!needToRun) {
+      // we have checked in `needToRunScript`
+      return previous!
+    }
+    const result = await script.transactionForDeployment(signer, params)
+    await signer.submitTransaction(result.unsignedTx)
+    const confirmed = await waitTxConfirmed(signer.provider, result.txId, network.confirmations)
+    const runScriptResult: RunScriptResult = {
+      fromGroup: result.fromGroup,
+      toGroup: result.toGroup,
+      txId: result.txId,
+      blockHash: confirmed.blockHash,
+      codeHash: codeHash,
+      attoAlphAmount: params.attoAlphAmount?.toString(),
+      tokens: tokens
+    }
+    runScriptResults.set(key, runScriptResult)
+    return runScriptResult
```

Comment:
Let's log something like the following. Hardhat should have better log messages.
a) Executing script "name"...
b) The execution of script "name" is skipped as it has been executed.

Same for contracts.

Reply:
Yes, hardhat uses the [debug](https://github.com/debug-js/debug) library, which can be controlled by the environment variables.